Libc/gen/ftw.3
.\"	$OpenBSD: ftw.3,v 1.4 2003/10/30 18:52:58 jmc Exp $
.\"
.\" Copyright (c) 2003 Todd C. Miller <Todd.Miller@courtesan.com>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\" Sponsored in part by the Defense Advanced Research Projects
.\" Agency (DARPA) and Air Force Research Laboratory, Air Force
.\" Materiel Command, USAF, under agreement number F39502-99-1-0512.
.\"
.Dd May 20, 2003
.Dt FTW 3
.Os
.Sh NAME
.Nm ftw, nftw
.Nd traverse (walk) a file tree
.Sh SYNOPSIS
.Fd #include <ftw.h>
.Ft int
.Fo ftw
.Fa "const char *path"
.Fa "int (*fn)(const char *, const struct stat *, int)"
.Fa "int maxfds"
.Fc
.Ft int
.Fo nftw
.Fa "const char *path"
.Fa "int (*fn)(const\ char\ *, const\ struct\ stat\ *, int, struct\ FTW\ *)"
.Fa "int maxfds"
.Fa "int flags"
.Fc
.Sh DESCRIPTION
.Bf -symbolic
These functions are provided for compatibility with legacy code.
New code should use the
.Xr fts 3
functions.
.Ef
.Pp
The
.Fn ftw
and
.Fn nftw
functions traverse (walk) the directory hierarchy rooted in
.Fa path .
For each object in the hierarchy, these functions call the function
pointed to by
.Fa fn .
The
.Fn ftw
function passes this function a pointer to a NUL-terminated string containing
the name of the object, a pointer to a stat structure corresponding to the
object, and an integer flag.
The
.Fn nftw
function passes the aforementioned arguments plus a pointer to a
.Dv FTW
structure as defined by
.Aq Pa ftw.h
(shown below):
.Bd -literal
struct FTW {
    int base;	/* offset of basename into pathname */
    int level;	/* directory depth relative to starting point */
};
.Ed
.Pp
Possible values for the flag passed to
.Fa fn
are:
.Bl -tag -width FTW_DNR
.It Dv FTW_F
A regular file.
.It Dv FTW_D
A directory being visited in pre-order.
.It Dv FTW_DNR
A directory which cannot be read.
The directory will not be descended into.
.It Dv FTW_DP
A directory being visited in post-order
.No ( Ns Fn nftw
only).
.It Dv FTW_NS
A file for which no
.Xr stat 2
information was available.
The contents of the stat structure are undefined.
.It Dv FTW_SL
A symbolic link.
.It Dv FTW_SLN
A symbolic link with a non-existent target
.No ( Ns Fn nftw
only).
.El
.Pp
The
.Fn ftw
function traverses the tree in pre-order.
That is, it processes the directory before the directory's contents.
.Pp
The
.Fa maxfds
argument specifies the maximum number of file descriptors
to keep open while traversing the tree.
It has no effect in this implementation.
.Pp
The
.Fn nftw
function has an additional
.Fa flags
argument with the following possible values:
.Bl -tag -width FTW_MOUNT
.It Dv FTW_PHYS
Physical walk, don't follow symbolic links.
.It Dv FTW_MOUNT
The walk will not cross a mount point.
.It FTW_DEPTH
Process directories in post-order.
Contents of a directory are visited before the directory itself.
By default,
.Fn nftw
traverses the tree in pre-order.
.It FTW_CHDIR
Change to a directory before reading it.
By default,
.Fn nftw
will change its starting directory.
The current working directory will be restored to its original value before
.Fn nftw
returns.
.El
.Sh RETURN VALUES
If the tree was traversed successfully, the
.Fn ftw
and
.Fn nftw
functions return 0.
If the function pointed to by
.Fa fn
returns a non-zero value,
.Fn ftw
and
.Fn nftw
will stop processing the tree and return the value from
.Fa fn .
Both functions return \-1 if an error is detected.
.Sh ERRORS
The
.Fn ftw
and
.Fn nftw
functions may fail and set
.Va errno
for any of the errors specified for the library functions
.Xr close 2 ,
.Xr open 2 ,
.Xr stat 2 ,
.Xr malloc 3 ,
.Xr opendir 3
and
.Xr readdir 3 .
If the
.Dv FTW_CHDIR
flag is set, the
.Fn nftw
function may fail and set
.Va errno
for any of the errors specified for
.Xr chdir 2 .
In addition, either function may fail and set
.Va errno
as follows:
.Bl -tag -width Er
.It Bq Er EINVAL
The
.Fa maxfds
argument is less than 1 or greater than
.Dv OPEN_MAX .
.El
.Sh LEGACY ERRORS
The
.Fn ftw
and
.Fn nftw
functions are far more tolerant of symlink cycles and are lax in reporting
errors while accessing the initial path.  When
.Fn nftw
is passed
.Dv FTW_MOUNT
it will pass the callback the mount point.
.Sh SEE ALSO
.Xr chdir 2 ,
.Xr close 2 ,
.Xr open 2 ,
.Xr stat 2 ,
.Xr compat 5 ,
.Xr fts 3 ,
.Xr malloc 3 ,
.Xr opendir 3 ,
.Xr readdir 3
.Sh STANDARDS
The
.Fn ftw
and
.Fn nftw
functions conform to
.St -p1003.1-2001 and
.St -susv3 .
.Sh HISTORY
Prior to MacOS X 10.4 
.Nm ftw
did not follow symlinks.
.Sh BUGS
The
.Fa maxfds
argument is currently ignored.


X11/ChangeLog

== 10.4 ChangeLog items ==

2005-03-09  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)

2004-10-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/InfoPlist.strings:
	removed CFBundleShortVersionString and CFBundleGetInfoString
	(3851089)

2004-10-19  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	added "Terminal" to the strings file	

	* xc/programs/Xserver/hw/apple/X11Application.m (prefs_get):
	added code to call NSLocalizedString on each name in the
	default applications menu

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessKnowsWindowNumber): added check for zero window id

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	Expanded backslash escape sequences since .strings files don't
	seem to handle them correctly. (3833239)

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/vfb/InitOutput.c: workaround 3834723
	by rearranging include order

2004-10-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Controller.m: QuartzRunClient
	function now has a second arg - whether or not to quote the
	command (3816789, 3572554)

2004-10-08  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/parse.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/Attrib.c: merged Xpm changes from XFree86
	HEAD branch (3744542, 3778302)

2004-09-29  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: fixed
	typo (3820725)

2004-09-16  John Harper  <jsh@apple.com>

	* xc/programs/xterm/main.c: call openpty() to scan for a pseudo
	terminal (3768572)

	* xc/programs/xterm/xterm.man: fix documentation of -ls option
	(3803120)

2004-09-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: nib
	changes for 3313133 and 3316468

2004-09-03  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: gave
	applications dialog window a minimum size

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	new file - ran genstrings on the .m files

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	added Localizable.strings

	* xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: added comments
	for translators

	* xc/programs/Xserver/hw/apple/Imakefile: set OBJROOT and
	SYMROOT when calling xcodebuild

2004-08-25  John Harper  <jsh@apple.com>

	* quartz-wm/x-window.m: don't grab Mod1-grave for window
	cycling, only Mod2 (3774967)

2004-08-23  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix typo in
	$XINITRC code (3621248)

2004-07-16  John Harper  <jsh@apple.com>

	* merged changes from XFree86 4.4.0 and fixed resulting
	brokenness (3570235, 3612446, 3548675, 3404805)

2004-06-10  John Harper  <jsh@apple.com>

	* xinitrc: quote shell expanded filenames to support home
	directories with spaces (3674992)

2004-03-29  John Harper  <jsh@apple.com>

	* FS/FSFontInfo.c, FS/FSFtNames.c, FS/FSGetCats.c,
	FS/FSListCats.c, FS/FSListExt.c, FS/FSOpenServ.c,
	FS/FSQGlyphs.c, FS/FSQXExt.c, FS/FSQXInfo.c, FS/FSlibint.h,
	FS/FSlibos.h, font/fc/fsconvert.c, font/fc/fserve.c,
	font/fc/fslibos.h, font/fontfile/dirfile.c,
	font/fontfile/encparse.c, font/fontfile/fontdir.c,
	font/fontfile/fontfile.c: merged font security fixes from
	XFree86 4.3 branch

2004-03-11  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.c (QuartzRunClient):
	added command quoting hack to fix 3572554

	* xc/programs/Xserver/hw/apple/bundle-main.c (define_named):
	fix for 3571829 (writing bad address size into xauth item)

2004-02-09  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c (start_client):
	respect value of $XINITRC. Also default to xinitrc in
	/etc/X11/xinit not the one under /usr/X11R6


ksh/ksh/src/lib/libast/comp/ftw.c
/*******************************************************************
*                                                                  *
*             This software is part of the ast package             *
*                Copyright (c) 1985-2004 AT&T Corp.                *
*        and it may only be used by you under license from         *
*                       AT&T Corp. ("AT&T")                        *
*         A copy of the Source Code Agreement is available         *
*                at the AT&T Internet web site URL                 *
*                                                                  *
*       http://www.research.att.com/sw/license/ast-open.html       *
*                                                                  *
*    If you have copied or used this software without agreeing     *
*        to the terms of the license you are infringing on         *
*           the license and copyright and are violating            *
*               AT&T's intellectual property rights.               *
*                                                                  *
*            Information and Software Systems Research             *
*                        AT&T Labs Research                        *
*                         Florham Park NJ                          *
*                                                                  *
*               Glenn Fowler <gsf@research.att.com>                *
*                David Korn <dgk@research.att.com>                 *
*                 Phong Vo <kpv@research.att.com>                  *
*                                                                  *
*******************************************************************/
#pragma prototyped
/*
 * ftw implementation
 */

#include <ast.h>
#include <ftw.h>

static int	(*ftw_userf)(const char*, const struct stat*, int);

static int
ftw_user(Ftw_t* ftw)
{
	register int	n = ftw->info;

	if (n & (FTW_C|FTW_NX))
		n = FTW_DNR;
	else if (n & FTW_SL)
		n = FTW_NS;
	return (*ftw_userf)(ftw->path, &ftw->statb, n);
}

int
ftw(const char* path, int(*userf)(const char*, const struct stat*, int), int depth)
{
	NoP(depth);
	ftw_userf = userf;
	return ftwalk(path, ftw_user, FTW_DOT, NiL);
}
